### PR TITLE
Improve README

### PR DESCRIPTION
**Pull request by pierreguilmin**
You can directly head to the branch to see how it looks like > https://github.com/dynamiqs/dynamiqs/tree/revamp-readme.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,24 +2,134 @@
...
 
+Here are the main features we're proud of:
+
+- Run your simulation on GPUs, to gain a significant speedup for large Hilbert space dimensions.
+- Batch your simulations to run them concurrently on multiple different Hamiltonians or initial states.
+- Explore quantum-specific solvers that preserve the properties of the state, such as positivity and trace.
+- Compute the gradient of any function that depends on the quantum state at any given time with respect to any parameter of the Hamiltonian, Lindblad jump operators, or initial state.
+- Choose to compute the gradient using fast automatic differentiation for smaller systems, or opt for the adjoint method for constant memory cost.
+- Use the library as a drop-in replacement for [QuTiP](https://qutip.org/) by directly passing QuTiP-defined quantum objects to our solvers.
+- Implement your own solvers with ease by subclassing our base solver class and focusing directly on the solver logic.
```
Comment (gautierronan) [inline README.md:24]:
```suggestion
- Running simulation on GPUs, with a significant speedup for large Hilbert space dimensions.
- Batching many simulations on multiple Hamiltonians or initial states to run them concurrently.
- Exploring quantum-specific solvers that preserve the properties of the state, such as trace and positivity.
- Computing gradients of any function of one or several quantum states with respect to any parametrized Hamiltonian, loss operator, or initial state.
- Using the library as a drop-in replacement for [QuTiP](https://qutip.org/) by directly passing QuTiP-defined quantum objects to our solvers.
- Implementing your own solvers with ease by subclassing our base solver class and focusing directly on the solver logic.
```
Some rewording suggestions, and also removed the point about the adjoint (I'd like to avoid any publicity until the paper is out).

File: README.md [hunk 2]
```diff
@@ -2,24 +2,134 @@
...
+- Enjoy reading our carefully crafted documentation on our website: <https://www.dynamiqs.org>.
+- Benefit from a well-structured library with a rigorous test suite, continuous integration and best coding practices (and a team of friendly enthusiast developers :nerd_face:).
```
Comment (gautierronan) [inline README.md:26]:
```suggestion
```
These ones are not so much related to the library, but rather they should be automatic for any well-written library. So I would keep them out. This makes it so readers can focus on the essentials a bit more also by having less points.

File: README.md [hunk 3]
```diff
@@ -2,24 +2,134 @@
...
+
+We hope that this library will prove beneficial to the community for simulations of large quantum systems like coupled bosonic modes, for gradient-based parameter estimation through experimental data fitting, and for efficient quantum optimal control.
+
```
Comment (gautierronan) [inline README.md:29]:
```suggestion
```
Not so sure about this sentence. I would be in favor of going straight to the point and keeping it out. But if you want we can keep it with some rewording, see below:
```suggestion
We hope that this library will prove beneficial to the community for e.g. simulations of large quantum systems, gradient-based parameter estimation, or large-scale quantum optimal control.
```

File: README.md [hunk 4]
```diff
@@ -2,24 +2,134 @@
...
+We will soon make a first release of the library on PyPi. In the meantime, you can clone the repository locally and install directly from source:
+
+```shell
+git clone https://github.com/dynamiqs/dynamiqs.git
+pip install -e dynamiqs
+```
+
+## Examples
+
+### Simulate a leaky bosonic mode
```
Comment (gautierronan) [inline README.md:41]:
```suggestion
### Simulate a lossy quantum harmonic oscillator
```

File: README.md [hunk 5]
```diff
@@ -2,24 +2,134 @@
...
+
+Let's simulate a leaky cavity with Hamiltonian $H=\Delta a^\dagger a$ and a single Lindblad jump operator $L=\sqrt{\kappa} a$, using QuTiP to define these operators:
```
Comment (gautierronan) [inline README.md:43]:
```suggestion
This first example shows simulation of a lossy harmonic oscillator with Hamiltonian $H=\Delta a^\dagger a$ and a single loss operator $L=\sqrt{\kappa} a$ using QuTiP-defined objects:
```

File: README.md [hunk 6]
```diff
@@ -2,24 +2,134 @@
...
+
+```python
+import dynamiqs as dq
+import numpy as np
+import qutip as qt
```
Comment (gautierronan) [inline README.md:48]:
```suggestion
import qutip as qt
import numpy as np
```

File: README.md [hunk 7]
```diff
@@ -2,24 +2,134 @@
...
+import torch
+
+# parameters
+n = 128
+delta = 1.0
+kappa = 0.5
```
Comment (gautierronan) [inline README.md:54]:
```suggestion
kappa = 0.1
```
(more usual to have very small loss compared to the oscillator frequency)

File: README.md [hunk 8]
```diff
@@ -2,24 +2,134 @@
...
+alpha0 = 1.0
+
+# QuTiP operators, initial state and saving times
+a = qt.destroy(n)
+H = a.dag() * a
+jump_ops = [np.sqrt(kappa) * a]
+rho0 = qt.coherent(n, alpha0)
```
Comment (gautierronan) [inline README.md:61]:
```suggestion
psi0 = qt.coherent(n, alpha0)
```

File: README.md [hunk 9]
```diff
@@ -2,24 +2,134 @@
...
+|██████████| 100.0% - time 00:00/00:00
+==== Result ====
+Method       : Dopri5
+Start        : 2023-07-06 22:52:30
+End          : 2023-07-06 22:52:31
+Total time   : 0.61 s
+states       : Tensor (101, 128, 128) | 12.62 Mb
+```
+
+### Compute the gradient with respect to some parameters
```
Comment (gautierronan) [inline README.md:84]:
```suggestion
### Compute gradients with respect to some parameters
```

File: README.md [hunk 10]
```diff
@@ -2,24 +2,134 @@
...
+n = 128
+delta = 1.0
+kappa = torch.tensor([0.5], requires_grad=True)
+alpha0 = torch.tensor([1.0], requires_grad=True)
+
+# dynamiqs operators, initial state and saving times
+a = dq.destroy(n)
+H = a.mH @ a
+jump_ops = [torch.sqrt(kappa) * a]
+rho0 = dq.coherent(n, alpha0)
```
Comment (gautierronan) [inline README.md:103]:
```suggestion
psi0 = dq.coherent(n, alpha0)
```

File: README.md [hunk 11]
```diff
@@ -2,24 +2,134 @@
...
+Output:
+
+```shell
+(tensor([-0.6065]), tensor([1.2131]))
+```
 
 ## Performance
 
-:construction: WIP
+While we have not yet conducted a rigorous benchmark of the library, we have some preliminary results. For instance, simulating two coupled dissipative bosonic modes each truncated at 32 photons takes approximately **2 minutes** with QuTiP, compared to **4 seconds** with dynamiqs (when executed on a standard GPU).
```
Comment (gautierronan) [inline README.md:127]: This comment is not super rigourous, I would keep it out for now. Once we have clearer benchmarks, we can add something similar back.

**Review by gautierronan — CHANGES_REQUESTED** [review]
Thanks a lot for this ! It's really useful for people to start considering the library seriously. Maybe we can copy paste this on the front-page of the documentation just for now?

Below are some comments, mostly some wording suggestions and a few things I would remove/modify. Otherwise it's great !

File: README.md [hunk 12]
```diff
@@ -2,24 +2,131 @@
...
+```shell
+git clone https://github.com/dynamiqs/dynamiqs.git
+pip install -e dynamiqs
+```
+
+## Examples
+
+### Simulate a lossy quantum harmonic oscillator
+
+This first example shows simulation of a lossy harmonic oscillator with Hamiltonian $H=\omega a^\dagger a$ and a single junp operator $L=\sqrt{\kappa} a$ using QuTiP-defined objects:
```
Comment (gautierronan) [inline README.md:40]:
```suggestion
This first example shows simulation of a lossy harmonic oscillator with Hamiltonian $H=\omega a^\dagger a$ and a single jump operator $L=\sqrt{\kappa} a$ using QuTiP-defined objects:
```

File: README.md [hunk 13]
```diff
@@ -2,24 +2,131 @@
...
+Method       : Dopri5
+Start        : 2023-07-07 10:26:13
+End          : 2023-07-07 10:26:13
+Total time   : 0.53 s
+states       : Tensor (101, 128, 128) | 12.62 Mb
+```
+
+### Compute gradients with respect to some parameters
+
+Suppose that in the above example, we want to compute gradients of the final state photon number expectation value $\text{loss}=\braket{a^\dagger a}=\mathrm{Tr}[a^\dagger a \rho]$ with respect to the decay rate $\kappa$ and the initial coherent state amplitude $\alpha_0$. For this computation, we will define the objects with dynamiqs:
```
Comment (gautierronan) [inline README.md:83]:
```suggestion
Suppose that in the above example, we want to compute the gradient of the number of photons in the final state, $\bar{n} = \mathrm{Tr}[a^\dagger a \rho(t_f)]$, with respect to the decay rate $\kappa$ and the initial coherent state amplitude $\alpha_0$. For this computation, we will define the objects with dynamiqs:
```

File: README.md [hunk 14]
```diff
@@ -2,24 +2,131 @@
...
+# simulation
+options = dq.options.Dopri5(gradient_alg='autograd', verbose=False)
+result = dq.mesolve(H, jump_ops, rho0, t_save, options=options)
+
+# gradient computation
+loss = dq.expect(a.mH @ a, result.states[-1])  # Tr[a^dag a rho]
+grads = torch.autograd.grad(loss, (kappa, alpha0))
+print(
+    f'gradient w.r.t to kappa : {grads[0]}\n'
+    f'gradient w.r.t to alpha0: {grads[1]}'
```
Comment (gautierronan) [inline README.md:115]:
```suggestion
    f'Gradient wrt to kappa  : {grads[0]}\n'
    f'Gradient wrt to alpha0 : {grads[1]}'
```

File: README.md [hunk 15]
```diff
@@ -2,24 +2,131 @@
...
+)
+```
 
-## Usage
+Output:
 
-:construction: WIP
+```shell
+gradient w.r.t to kappa : tensor([-0.9048])
+gradient w.r.t to alpha0: tensor([1.8097])
```
Comment (gautierronan) [inline README.md:123]:
```suggestion
Gradient wrt to kappa  : tensor([-0.9048])
Gradient wrt to alpha0 : tensor([1.8097])
```

**Review by gautierronan — APPROVED** [review]
